source/slang/slang-ir-use-uninitialized-values.cpp

```diff
@@ -348,8 +348,135 @@ namespace Slang
+                return true;
+        }
+
+        return false;
+    }
+
+    static IRInst* traceInstOrigin(IRInst* inst)
+    {
+        if (auto load = as<IRLoad>(inst))
+            return traceInstOrigin(load->getPtr());
+
+        return inst;
+    }
+
+    static bool isFieldUsed(IRStructField* field, const List<IRStructKey*>& used)
+    {
+        for (auto key : used)
+        {
+            if (field->getKey() == key)
+                return true;
+        }
+
+        return false;
+    }
+
+    static List<IRStructField*> checkFieldsFromExit(ReachabilityContext& reachability, IRReturn* ret, IRStructType* type)
+    {
+        IRInst* origin = traceInstOrigin(ret->getVal());
+
+        // We don't want to warn on delegated construction
+        if (!isUninitializedValue(origin))
+            return {};
+
+        // Now we can look for all references to fields
+        List<IRStructKey*> usedKeys;
+        for (auto use = origin->firstUse; use; use = use->nextUse)
+        {
+            IRInst* user = use->getUser();
+            
+            auto fieldAddress = as<IRFieldAddress>(user);
+            if (!fieldAddress || !isInstStoredInto(reachability, ret, user))
+                continue;
+
+            IRInst* field = fieldAddress->getField();
+            usedKeys.add(as<IRStructKey>(field));
+        }
+
+        List<IRStructField*> uninitializedFields;
+
+        auto fields = type->getFields();
+        for (auto field : fields)
+        {
+            if (canIgnoreType(field->getFieldType(), nullptr))
+                continue;
+
+            if (!isFieldUsed(field, usedKeys))
+                uninitializedFields.add(field);
+        }
+        
+        return uninitializedFields;
+    }
+
+    static void checkConstructor(IRFunc* func, ReachabilityContext& reachability, DiagnosticSink* sink)
+    {
+        IRStructType* stype = as<IRStructType>(func->getResultType());
+        if (!stype)
+            return;
+        
+        auto synthesized = func->findDecoration<IRSynthesizedDecoration>();
+        
+        auto emitWarnings = [&](const List<IRStructField*>& fields, IRReturn* ret)
+        {
+            for (auto field : fields)
+            {
+                if (synthesized)
+                {
+                    sink->diagnose(field->getKey(),
+                        Diagnostics::fieldNotDefaultInitialized,
+                        stype,
+                        field->getKey());
+                }
+                else
+                {
+                    sink->diagnose(ret,
+                        Diagnostics::constructorUninitializedField,
+                        field->getKey());
+                }
+            }
+
+        };
+
+        // Work backwards, get exit points and find sources
+        for (auto block : func->getBlocks())
+        {
+            for (auto inst = block->getFirstInst(); inst; inst = inst->next)
+            {
+                auto ret = as<IRReturn>(inst);
```

Comment:
You may want to use `TerminatorInst` like
```
auto ret = as<TerminatorInst>(inst);
```

Reply:
We need an explicit return so we can get the value returned (constructed) and trace is back so a `IRTerminatorInst` wouldn't be enough.

Reply:
There are other ways to return from the middle of the constructor.
`__intrinstic_asm` is one of them.
And they are grouped into `IRTerminatorInst`.